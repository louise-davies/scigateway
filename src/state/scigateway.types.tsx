import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadedAuthType = 'scigateway:loaded_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadScheduledMaintenanceStateType =
  'scigateway:load_scheduled_maintenance_state';
export const LoadMaintenanceStateType = 'scigateway:load_maintenance_state';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SendThemeOptionsType = 'scigateway:api:send_themeoptions';
export const LoadDarkModePreferenceType =
  'scigateway:load_dark_mode_preference';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';
export const RegisterStartUrlType = 'scigateway:register_start_url';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
  showHelpPageButton: boolean;
}

export interface StartUrlPayload {
  startUrl: string;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface SendThemeOptionsPayload {
  theme: Theme;
  broadcast: boolean;
}

export interface LoadDarkModePreferencePayload {
  darkMode: boolean;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface ScheduledMaintenanceStatePayLoad {
  scheduledMaintenance: ScheduledMaintenanceState;
}

export interface ScheduledMaintenanceState {
  show: boolean;
  message: string;
}

export interface MaintenanceStatePayLoad {
  maintenance: MaintenanceState;
}

export interface MaintenanceState {
  show: boolean;
  message: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
