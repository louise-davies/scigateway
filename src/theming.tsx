import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#FF6900',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  primary: string;
  secondary: string;
  background: string;
  paper: string;
  blue: string;
  orange: string;
  red: string;
  grey: string;
}

const DARK_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

//For experimenting
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const DARK_MODE_COLOURS_HIGH_CONTRAST: ThemeColours = {
  primary: '#86B4FF',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#003088',
  background: '#FAFAFA',
  paper: '#FFF',
  blue: '#003088',
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
};

export const buildTheme = (darkModePreference: boolean): Theme => {
  let options: UKRIThemeOptions;
  const colours = darkModePreference ? DARK_MODE_COLOURS : LIGHT_MODE_COLOURS;

  const overrides = {
    MuiLink: {
      root: {
        color: colours.blue,
      },
    },
    MuiTabs: {
      indicator: {
        color: STATIC_COLOURS.darkBlue,
        textDecoration: 'underline',
      },
    },
    MuiFormLabel: {
      root: {
        '&$error': {
          color: colours.red,
        },
        '&$focused': {
          color: colours.blue,
        },
      },
      asterisk: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiBadge: {
      colorPrimary: {
        backgroundColor: STATIC_COLOURS.orange,
      },
    },
    MuiInput: {
      underline: {
        '&$error:after': {
          borderBottomColor: colours.red,
        },
        '&:after': {
          borderBottomColor: colours.blue,
        },
      },
    },
    MuiOutlinedInput: {
      root: {
        '&$error $notchedOutline': {
          borderColor: colours.red,
        },
      },
    },
    MuiFormHelperText: {
      root: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiPickersToolbar: {
      toolbar: {
        backgroundColor: STATIC_COLOURS.darkBlue,
      },
    },
    MuiPickersCalendarHeader: {
      dayLabel: {
        color: colours.grey,
      },
    },
    MuiPickersDay: {
      current: {
        color: colours.blue,
      },
      dayDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersYear: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      yearSelected: {
        color: colours.blue,
      },
      yearDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersMonth: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      monthSelected: {
        color: colours.blue,
      },
      monthDisabled: {
        color: colours.grey,
      },
    },
  };

  if (darkModePreference) {
    options = {
      palette: {
        // Light/dark mode
        type: 'dark',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#257fff',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  } else {
    options = {
      palette: {
        // Light/dark mode
        type: 'light',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#1E5DF8',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  }

  return createMuiTheme(options);
};

function mapThemeProviderStateToProps(
  state: StateType
): { prefersDarkMode: boolean } {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference: boolean = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  return (
    <MuiThemeProvider theme={buildTheme(darkModePreference)}>
      {props.children}
    </MuiThemeProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
