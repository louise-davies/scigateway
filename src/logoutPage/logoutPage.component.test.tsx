import React from 'react';
import LogoutPage, {
  UnconnectedLogoutPage,
  CombinedLogoutPageProps,
} from './logoutPage.component';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { authState, initialState } from '../state/reducers/scigateway.reducer';
import { Provider } from 'react-redux';
import { push } from 'connected-react-router';
import thunk from 'redux-thunk';
import TestAuthProvider from '../authentication/testAuthProvider';
import { buildTheme } from '../theming';
import { ThemeProvider, StyledEngineProvider } from '@mui/material/styles';
import { createLocation } from 'history';
import { mount, shallow } from 'enzyme';

describe('logout page component', () => {
  let props: CombinedLogoutPageProps;
  let mockStore;
  let state: StateType;

  beforeEach(() => {
    mockStore = configureStore([thunk]);
    state = {
      scigateway: { ...initialState, authorisation: { ...authState } },
      router: { location: createLocation('/') },
    };

    state.scigateway.authorisation.provider = new TestAuthProvider(
      'test-token'
    );
  });

  const theme = buildTheme(false);

  it('renders the logout page correctly with default avatar ', () => {
    props = {
      user: {
        username: 'simple/root',
        isAdmin: false,
        avatarUrl: '',
      },
      res: undefined,
    };
    const wrapper = shallow(<UnconnectedLogoutPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the logout page correctly with avatar using avatarurl) ', () => {
    props = {
      user: {
        username: 'simple/root',
        isAdmin: false,
        avatarUrl: 'test_url',
      },
      res: undefined,
    };

    const wrapper = shallow(<UnconnectedLogoutPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('signs out if sign out clicked', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <LogoutPage />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper
      .find('[data-test-id="logout-page-button"]')
      .last()
      .simulate('click');

    expect(testStore.getActions().length).toEqual(2);
    expect(testStore.getActions()[0]).toEqual({ type: 'scigateway:signout' });
    expect(testStore.getActions()[1]).toEqual(push('/'));
  });
});
