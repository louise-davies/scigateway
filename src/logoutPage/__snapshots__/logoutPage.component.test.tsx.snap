// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logout page component renders the logout page correctly with avatar using avatarurl)  1`] = `
<div
  className="logout-page"
>
  <ForwardRef(Paper)
    sx={[Function]}
  >
    <ForwardRef(Avatar)
      alt="user"
      src="test_url"
      sx={
        Object {
          "margin": 1,
        }
      }
    />
    <ForwardRef(Typography)
      sx={
        Object {
          "marginTop": 1,
        }
      }
    >
      username-description
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "fontSize": "17px",
          "fontWeight": "bold",
          "paddingTop": "3px",
        }
      }
    >
      simple/root
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "color": "secondary.main",
          "marginTop": 1,
        }
      }
    >
      logout-message
    </ForwardRef(Typography)>
    <ForwardRef(Button)
      color="primary"
      data-test-id="logout-page-button"
      onClick={[Function]}
      sx={
        Object {
          "marginTop": 1,
        }
      }
      variant="contained"
    >
      <ForwardRef(Typography)
        color="inherit"
        noWrap={true}
        sx={
          Object {
            "marginTop": "3px",
          }
        }
      >
        logout-button
      </ForwardRef(Typography)>
    </ForwardRef(Button)>
  </ForwardRef(Paper)>
</div>
`;

exports[`logout page component renders the logout page correctly with default avatar  1`] = `
<div
  className="logout-page"
>
  <ForwardRef(Paper)
    sx={[Function]}
  >
    <ForwardRef(Avatar)
      sx={
        Object {
          "backgroundColor": [Function],
          "color": "#FFFFFF",
          "margin": 1,
        }
      }
    >
      <Memo(ForwardRef(AccountCircleIcon)) />
    </ForwardRef(Avatar)>
    <ForwardRef(Typography)
      sx={
        Object {
          "marginTop": 1,
        }
      }
    >
      username-description
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "fontSize": "17px",
          "fontWeight": "bold",
          "paddingTop": "3px",
        }
      }
    >
      simple/root
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "color": "secondary.main",
          "marginTop": 1,
        }
      }
    >
      logout-message
    </ForwardRef(Typography)>
    <ForwardRef(Button)
      color="primary"
      data-test-id="logout-page-button"
      onClick={[Function]}
      sx={
        Object {
          "marginTop": 1,
        }
      }
      variant="contained"
    >
      <ForwardRef(Typography)
        color="inherit"
        noWrap={true}
        sx={
          Object {
            "marginTop": "3px",
          }
        }
      >
        logout-button
      </ForwardRef(Typography)>
    </ForwardRef(Button)>
  </ForwardRef(Paper)>
</div>
`;
